### PR TITLE
[SOLID] Skip property that is being changed by func call

### LINKED ISSUE #3810: ChangeReadOnlyPropertyWithDefaultValueToConstantRector wrongly creates variables and constants
# Bug Report

| Subject        | Details                                                         |
| :------------- | :---------------------------------------------------------------|
| Rector version | v0.7.54              |
| Installed as   | composer dependency                      |

```diff
    ---------- begin diff ----------
--- Original
+++ New
@@ -22,7 +22,7 @@
  */
 final class Validator
 {
-    private $schemas = ['1.2'];
+    private const SCHEMAS = ['1.2'];
     /** @var array */
     private $errors = [];

@@ -97,7 +97,8 @@
      */
     private function getSchema(?string $schemaVersion): string
     {
-        $schemaVersion = $schemaVersion ?? end($this->schemas);
+        $self::SCHEMAS = self::SCHEMAS;
+        $schemaVersion = $schemaVersion ?? \end($self::SCHEMAS);
         $finder        = new Finder();
         $iterator      = $finder->name($schemaVersion . '.xsd')->in(__DIR__ . DIRECTORY_SEPARATOR . '../Schemas')->getIterator();
         $iterator->rewind();
    ----------- end diff -----------
```

~~Caused by: `Rector\SOLID\Rector\Property\ChangeReadOnlyPropertyWithDefaultValueToConstantRector`~~

Applied rules:

 * Rector\Core\Rector\Argument\ArgumentAdderRector
 * Rector\Core\Rector\Argument\ArgumentDefaultValueReplacerRector
 * Rector\Core\Rector\Argument\ArgumentRemoverRector
 * Rector\Core\Rector\Argument\SwapFuncCallArgumentsRector
 * Rector\Core\Rector\ClassMethod\WrapReturnRector
 * Rector\Core\Rector\FuncCall\RemoveFuncCallArgRector
 * Rector\MysqlToMysqli\Rector\FuncCall\MysqlFuncCallToMysqliRector
 * Rector\MysqlToMysqli\Rector\FuncCall\MysqlQueryMysqlErrorWithLinkRector
 * Rector\Php54\Rector\FuncCall\RemoveReferenceFromCallRector
 * Rector\Php70\Rector\FuncCall\NonVariableToVariableOnFunctionCallRector
 * Rector\Php70\Rector\FuncCall\RandomFunctionRector
 * Rector\Renaming\Rector\Annotation\RenameAnnotationRector
 * Rector\Renaming\Rector\Function_\RenameFunctionRector
 * Rector\SOLID\Rector\Property\ChangeReadOnlyPropertyWithDefaultValueToConstantRector

The proposed change is not correct.

PHP (via PHPStorm) complains about it:

> Only variables can be passed by reference

## Minimal PHP Code Causing Issue

Demo (not working due to the yaml config): https://getrector.org/demo/b5fb1b21-4008-44aa-9a13-f34b93e95060
## Expected Behaviour

There are various alternatives:

1. Assign the value of the constant to a variable correctly named
2. Don't change the array into a constant (I suppose really really complex to achieve)

In case one, the diff should be this (assuming the rule is already applied):

```diff
     private function getSchema(?string $schemaVersion): string
     {
-        $self::SCHEMAS = self::SCHEMAS;
+        $schemas = self::SCHEMAS;
-        $schemaVersion = $schemaVersion ?? \end($self::SCHEMAS);
+        $schemaVersion = $schemaVersion ?? \end($schemas);
         $finder        = new Finder();
         $iterator      = $finder->name($schemaVersion . '.xsd')->in(__DIR__ . DIRECTORY_SEPARATOR . '../Schemas')->getIterator();
         $iterator->rewind();
```

As you can see, in fact, the generated variable `$self::SCHEMAS` is anyway not correct syntactically.

### PR DESCRIPTION
**Pull request by TomasVotruba**
Closes #3810

### CODE REVIEW

File: src/PhpParser/Node/Manipulator/PropertyManipulator.php [hunk 1]
```diff
@@ -32,6 +33,11 @@ final class PropertyManipulator
 {
     use DoctrineTrait;
 
+    /**
+     * @var string[]
+     */
+    private const MODIFIED_FUNCTION_NAMES = ['end'];
```
Comment (TomasVotruba) [inline src/PhpParser/Node/Manipulator/PropertyManipulator.php:39]: @Aerendir Could you think of other functions that change passed argument?

Comment (Aerendir) [reply src/PhpParser/Node/Manipulator/PropertyManipulator.php:39]: I'm not sure I'm understanding what you mean by "change passed argument".

But, if the example is `end()` function, then there are a lot of functions related to arrays that change the passed argument.

The ones that comes to my mind:

- https://www.php.net/manual/en/function.array-change-key-case.php
- https://www.php.net/manual/en/function.array-flip.php
- https://www.php.net/manual/en/function.array-pop.php
...

Do you mean this kind of functions?

Comment (TomasVotruba) [reply src/PhpParser/Node/Manipulator/PropertyManipulator.php:39]: Yes, exactly

Comment (TomasVotruba) [reply src/PhpParser/Node/Manipulator/PropertyManipulator.php:39]: Could you send PR to add them after this is merged?

Comment (Aerendir) [reply src/PhpParser/Node/Manipulator/PropertyManipulator.php:39]: For the moment I don't think will find the time... :(

I'm upgrading all my libraries and it is a tedious and long process... 😩

I should send a PR to Mautic and I'm not able to find 10 minutes 😓

Comment (TomasVotruba) [reply src/PhpParser/Node/Manipulator/PropertyManipulator.php:39]: Sure :+1: 

Whenever you're ready (or the rule fails again :D)

**Pull request merged**